WordPress/openverse
# Make usages of Redis resilient to absence of Redis

## Fixes



Fixes #3385 by @AetherUnbound 

## Description



### General changes

- Refactored fixtures to remove duplicates and set up a separate fixtures package.
- Added a factory for the `ContentProvider` model.
- Moved `CACHES` to a separate settings file.
- Replaced direct cache manipulations with fake Redis.

### Throttling

- Removed exempt IPs flow.
- Added handling for Redis absence.
	- All incoming requests will become allowed.
- Updated tests to account for headers being absent when history cannot be determined from cache.

### Search controller

- Added handling for Redis absence in filtered providers.
	- Filtered providers are computed from the DB.
- Added handling for Redis absence in stats.
	- Sources will be obtained from Elasticsearch.
	- Also fixed a bug in handling of outdated formats.
- Added tests for filtered providers computed from DB and sources computed from ES.

### OAuth2

- Added handling for Redis absence when checking usage.
	- Usage is shown as `None`.
	- Status code for the response is 424.
- Added tests for 424 response and `None` usage counts.

### Tallies

- Added handling for Redis absence when tallying providers.
	- Just log it!
- Added tests for the added log lines.

### Image proxy

- Converted multiple `incr` requests to a pipeline.
- Added handling for Redis absence when tallying response codes and HTTP errors and other errors.
	- Just log it!
- Added handling for Redis absence when determining image extension.
	- Extension never cached, always treated as unknown i.e. `None` and determined from `Content-Type` header.
- An `itertools.count` counter is used to space out exceptions being sent Sentry. Afaik, `itertools.count` is per worker process so we will still likely be sending 20x the number of Sentry event we expect.
- Added tests for the added log lines.

### Dead links

- Added handling for Redis absence when getting and saving query masks.
	- Query mask never cached, always starts with blank list i.e `[]` and recomputed.
- Added handling for Redis absence when checking image liveness.
	- All URLs will always be revalidated by checking HTTP response status code.
- Added tests for the recomputed query mask and log lines.

## Testing Instructions


All changes have accompanying CI tests. Additionally, you can try to use the API with the Redis service stopped and see that things still work normally.

## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
